### ISSUE TITLE
redefine 1D `self.N`, deprecate`self.J_pad`?

### ISSUE THREAD

**Issue by lostanlen** [kymatio/kymatio#857]
in 1D, `self.N` is the unpadded length of the input signal. It is defined in `ScatteringBase1D` as some derivative of `self.shape` and isn't being used anywhere else. 

Same for 2D where `self.shape` unpacks as `self.M`, `self.N`. 

I see no reason to keep these values as attributes in the object when:
* we have `self.shape` for that
* these values are only being used transiently, in `self.build()`
* our filterbank construction relies on `N_padded`, not `N`
* these values are trivially available in the forward via `self.N = len(x)`

In the spirit of preparing for scattering on variable-length inputs (`shape=None`, a long-requested feature !), i ask the permission to remove these attributes from `self`.

I am open to making a deprecation cycle where we convert these attributes to read-only properties and print a `DeprecationWarning` in v0.3 if anyone tries to access these numbers (NB: none of our examples access `self.N` and friends)

@janden @eickenberg

**Comment by lostanlen** [comment]
or else we decide that `self.N` now stands for `N_padded`, i.e., the length of the signal after padding = the length of the filters at the full resolution. This would actually be most consistent with the 3D module. But that implies to rename fields in the 2D module: `N_padded` becomes `N`, `M_padded` becomes `M`.

**Comment by janden** [comment]
I'd be in favor of removing. We could argue that these were intended to be private attributes and users shouldn't rely on them. Safe thing would be to issue a deprecation warning, however.

**Comment by lostanlen** [comment]
> We could argue that these were intended to be private attributes

@janden i came to the same conclusion. 


>  Safe thing would be to issue a deprecation warning, however.

i can do that !

**Comment by janden** [comment]
@eickenberg Your call on this API-breaking change.

**Comment by lostanlen** [comment]
this PR is related to JTFS. Here's my rationale:
* i'd like to avoid having to write a `jtfs_filter_factory`
* rather, i'd like to call `scattering_filter_factory` twice: once for the temporal filters and again (with different arguments) for the frequential filters
* but i don't want to have to assume that the input  of the frequential scattering has a power-of-two support. That would require a lot of padding (up to ~JQ zeros, i.e. >100) and thus woul hurt speed and memory usage. Rather i'd like to pad to a multiple of log2_F (aka the frequential "stride" )
* yet `scattering_filter_factory` currently takes `J_pad` (the log2 of the padded support) as input. Since J_pad must be integer, it follows that the padded input must be a power of two in length.
* so i will propose (in a separate PR) that `scattering_filter_factory` should take N_padded, not J_pad, as input.
* that future PR will improve consistency with Scattering2D, whose `scattering_filter_factory` requires takes `N_padded`, not its log2, as input.
* besides JTFS, that future PR will also benefit 1D (aka "plain time scattering") because it will speed up the code (smaller input) and reduce memory usage (better data/padding ratio).
* of course this PR will require its own test suite and careful review
* but for now, i think it is clear that J_pad is not necessary to instantiate the 1D scattering filter factory, so i propose to do it now (in 0.3) with the plan of removing it in 0.4.
* The removal will be justified by a concern of "principle of minimum surprise" between time scattering parameters (N,J,Q,T) and frequency scattering (N_fr, J_fr, Q_fr, T_fr). There won't be an integer named J_pad_fr
* in conclusion, i believe that this deprecation / API change is a reasonable price to pay in front of improved consistency and flexibility

@eickenberg i hope this clarifies

**Comment by janden** [comment]
So you're saying that we should allow the filters to have non-power-of-two sizes? Then yes, I agree with that. The filters are technically part of the public API, but still in a sort of grey zone (I don't believe they're documented). So I don't think it's too much of an issue to break it.

**Comment by janden** [comment]
But why even keep the `N` attribute? The padded size is, in a sense, an internal parameter for the implementation. Why do we need to expose it to the user at all?

**Comment by lostanlen** [comment]
Removing `self.N` entirely is not entirely trivial since `create_filters(self)` needs it in `self` to call `scattering_filter_factory` with `self.N` as first argument.
But i could easily rename it to `self._N` so that it does not appear to the user in `self.+TAB`

If we change our minds later and think that `self.N` (by which i mean, the padded N, not the input length `shape[0]`) should be made public, we could do:

```python
@property
def N(self):
    return len(self.phi_f[0])
```

in this way, the integer in `self._N` will only serve to carry across the output of `compute_temporal_support` (called in `self.build()`) over to the input of `scattering_filter_factory` (called in `self.create_filters()`)

**Comment by janden** [comment]
I think that sounds like a good idea.

**Comment by janden** [comment]
Might be good to rename it to something like `_N_padded`, though to make clear it's the padded size that's intended.
